### PR TITLE
Feature/machine learning

### PR DESCRIPTION
**Pull request by mahbx**
# Changes :gem:
add features to the machine learning module
* add function to classify images by binary data
* add function to delete the oldest iteration if there exists 10 or more iterations
  * this is needed because custom vision free version only allows us to have 10 iterations
* documenting a possible issue where two processes might have issues with synchronization if you have extremely long lived processes

### CODE REVIEW

File: machine_learning_utilities/CV_classifier.py [hunk 1]
```diff
@@ -78,6 +78,28 @@ def predict(self, img_url: str) -> Dict[str, float]:
...
+        res = self.predictor.classify_image(
+            self.project_id, self.iteration_name, png_img
+        )
+
+        pred_kv = dict([(i.tag_name, i.probability) for i in res.predictions])
+
+        return pred_kv
+
     def __chunks(self, lst, n):
         """Helper method used by upload_images() to upload URL chunks of 64, which is maximum chunk size in Azure Custom Vision."""
```
Comment (andreacx) [inline machine_learning_utilities/CV_classifier.py:104]: Fix formatting for this docstring!!!!!

File: machine_learning_utilities/CV_classifier.py [hunk 2]
```diff
@@ -152,25 +177,50 @@ def upload_images(self, labels: List) -> None:
...
+            Custom Vision allows maximum 10 iterations in the free version. 
+        """
+
+        iterations = self.trainer.get_iterations(self.project_id)
+
+        if len(iterations) >= 10:
+
+            iterations.sort(key=lambda i: i.created)
+            oldest_iteration = iterations[0].id
+            #print(oldest_iteration, iterations[0])
```
Comment (joaaax) [inline machine_learning_utilities/CV_classifier.py:192]: Do we need this?

File: machine_learning_utilities/CV_classifier.py [hunk 3]
```diff
@@ -196,27 +246,35 @@ def train(self, labels: list) -> None:
...
     blob_service_client = BlobServiceClient.from_connection_string(connect_str)
 
     test_url = "https://originaldataset.blob.core.windows.net/ambulance/4504435055132672.png"
+
     labels = ["ambulance", "bench", "circle", "star", "square"]
     classifier = CVClassifier(blob_service_client)
-    classifier.upload_images(labels)
+    # classifier.upload_images(labels)
+
+    # classifier.train(labels)
```
Comment (joaaax) [inline machine_learning_utilities/CV_classifier.py:267]: Do we need these?

Comment (mahbx) [reply machine_learning_utilities/CV_classifier.py:261]: yes we need these

**Pull request merged**